### ISSUE TITLE
Reproducing model zoos simCLR top 1 accuracy

### ISSUE THREAD

**Issue by spurra** [facebookresearch/vissl#315]
I'm attempting to reproduce the numbers for SimCLR as mentioned in the [model zoo's README](https://github.com/facebookresearch/vissl/blob/master/MODEL_ZOO.md#SimCLR) with the help of the pre-trained models.

VISSL seems to be fully functioning on my machine (i.e I can run training scripts). The steps I've taken are as follows.:
1) Downloaded ImageNet-2012 train and validation from https://image-net.org/challenges/LSVRC/2012/2012-downloads.php
2) Ran [this script](https://raw.githubusercontent.com/soumith/imagenetloader.torch/master/valprep.sh) to structure the validation set according to [VISSL documentation](https://vissl.readthedocs.io/en/latest/vissl_modules/data.html#expected-dataset-structure-for-imagenet-places205-places365). This is because the downloaded validation images don't come with subfolders like the training set does, for some reason.
3) I downloaded the "RN50 - 100 epoch" model from [here](https://dl.fbaipublicfiles.com/vissl/model_zoo/simclr_rn50_100ep_simclr_8node_resnet_16_07_20.8edb093e/model_final_checkpoint_phase99.torch) and put in in a folder called "model_zoo"
4) The model zoo for SimCLR points to [this json file](https://github.com/facebookresearch/vissl/blob/master/configs/config/model_zoo/benchmark_in1k_linear_simclr.json) to reproduce the numbers. I'm not sure how to run this with VISSL, so I decided to manually launch a training run via the following configuration: benchmark/linear_image_classification/imagenet1k/eval_resnet_8gpu_transfer_in1k_linear.yaml
5) I only have one GPU available, so I launch the following command: 
`python tools/run_distributed_engines.py config=benchmark/linear_image_classification/imagenet1k/eval_resnet_8gpu_transfer_in1k_linear.yaml config.DISTRIBUTED_NUM_PROC_PER_NODE=1 config.MODEL.SYNC_BN_CONFIG.SYNC_BN_TYPE=pytorch`

This seems to run fine. My question here is am I doing this correctly and will I achieve a top-1 accuracy close to the 64.4 as reported in the table?

**Comment by prigoyal** [comment]
Hi @spurra , thank you for reaching out. Yes, I would expect that you reproduce the number. The important thing is to ensure the learning rate is scaled properly as #gpus are changed. For this in VISSL, the https://github.com/facebookresearch/vissl/blob/master/configs/config/benchmark/linear_image_classification/imagenet1k/eval_resnet_8gpu_transfer_in1k_linear.yaml#L103 is provided to automatically adjust LR

**Comment by spurra** [comment]
Thanks for the response @prigoyal ! I'll let you know what numbers I get

**Comment by spurra** [comment]
I finished running the experiment. This is the output of the log hook:

`INFO 2021-05-12 09:38:02,635 log_hooks.py: 446: Rank: 0, name: test_accuracy_list_meter, value: {'top_1': {'conv1': 14.628, 'res2': 28.27, 'res3': 39.088, 'res4': 56.391999999999996, 'res5': 62.246}, 'top_5': {'conv1': 29.92, 'res2': 48.455999999999996, 'res3': 61.07, 'res4': 78.32000000000001, 'res5': 85.15599999999999}}
`
I assume the top 1 accuracy of res5 is relevant, I'm not quite sure what the others indicate. I achieve 62.246 vs 64.4 which was reported for the RN50 model trained for 100 epochs. For completeness, I'm uploading the train_config.yaml file which was produced by the code which can be found here: https://gist.github.com/spurra/d5b89caccbd614522eb19e6bc3a9e2d9

Is this performance discrepancy within expected range?

**Comment by prigoyal** [comment]
thank you for reporting @spurra , let me try to rerun the benchmark and come back to you. We should be able to repro exactly. I'll try the original setting and your setting as well.

**Comment by prigoyal** [comment]
> EDIT2: I'm a little confused by your comment regarding adjusting batch size as #gpus are changed. You state that https://github.com/facebookresearch/vissl/blob/master/configs/config/benchmark/linear_image_classification/imagenet1k/eval_resnet_8gpu_transfer_in1k_linear.yaml#L103 does this in VISSL. However the note in the documentation [here](https://vissl.readthedocs.io/en/latest/ssl_approaches/simclr.html#vary-the-number-of-gpus) suggests that I have to do this manually. Also, the documentation on the configs found [here](https://github.com/facebookresearch/vissl/blob/master/vissl/config/defaults.yaml#L948) suggest this just applies the linear scaling rule. If I understand correctly, this just scales the learning rate according to the batch size, not the number of GPUs. Could you please clarify if I need to normalize the SimCLR loss by the total batch size or this is already taken care of? Thanks!

correct. Only learning rate will be scaled. Not the batch size. You should keep the batch size per gpu consistent and then learning rate will be auto-scaled.

**Comment by prigoyal** [comment]
Based on the responses above and the report from @doulemint , it looks like this benchmark reproduces numbers . Please free to reopen the task if that's not the case still.

Also, as a follow-up, we will look into providing the clarification for the learning rate scaling feature in VISSL in our docs, code comments etc. :)

**Comment by iseessel** [comment]
@prigoyal Sorry to cross wires here -- I've actually been looking into this -- see below and lmk if you have anything to add. 

Hi @spurra + @doulemint,

I was able to reproduce your 1-gpu numbers, as well as our reported 8-gpu numbers. See below for the full results. These represent the best reported accuracy for train/test for each layer. As a side note, I believe the lr appears to be zero because we are rounding before logging to tensroboard: https://github.com/facebookresearch/vissl/blob/main/vissl/hooks/tensorboard_hook.py#L265

```
1GPU: 
rn50_in1k_simclr_100ep_eval_resnet_8gpu_transfer_in1k_linear_eval_resnet_1gpu_transfer_in1k_linear_14_10_21
None [ rn50_in1k_simclr_100ep_eval_resnet_8gpu_transfer_in1k_linear_eval_resnet_1gpu_transfer_in1k_linear_14_10_21 ] :
 - train.top_1.conv1 : 0.105445 (54)
 - train.top_1.res2 : 0.224741 (54)
 - train.top_1.res3 : 0.32512 (50)
 - train.top_1.res4 : 0.519181 (50)
 - train.top_1.res5 : 0.628983 (50)
 - train.top_5.conv1 : 0.228618 (50)
 - train.top_5.res2 : 0.4013169999999999 (50)
 - train.top_5.res3 : 0.524946 (54)
 - train.top_5.res4 : 0.729936 (50)
 - train.top_5.res5 : 0.843128 (52)
 - test.top_1.conv1 : 0.14676 (51)
 - test.top_1.res2 : 0.28148 (51)
 - test.top_1.res3 : 0.39164 (43)
 - test.top_1.res4 : 0.56474 (47)
 **- test.top_1.res5 : 0.62368 (47)**
 - test.top_5.conv1 : 0.29776 (51)
 - test.top_5.res2 : 0.48412 (51)
 - test.top_5.res3 : 0.6116 (47)
 - test.top_5.res4 : 0.78462 (47)
 - test.top_5.res5 : 0.85202 (43)

8 GPU: 
rn50_in1k_simclr_100ep_eval_resnet_8gpu_transfer_in1k_linear_eval_resnet_8gpu_transfer_in1k_linear_14_10_21
None [ rn50_in1k_simclr_100ep_eval_resnet_8gpu_transfer_in1k_linear_eval_resnet_8gpu_transfer_in1k_linear_14_10_21 ] :
 - train.top_1.conv1 : 0.109691 (48)
 - train.top_1.res2 : 0.235089 (52)
 - train.top_1.res3 : 0.337494 (52)
 - train.top_1.res4 : 0.536332 (52)
 - train.top_1.res5 : 0.652946 (52)
 - train.top_5.conv1 : 0.23703700000000003 (48)
 - train.top_5.res2 : 0.41495 (52)
 - train.top_5.res3 : 0.539311 (52)
 - train.top_5.res4 : 0.744265 (52)
 - train.top_5.res5 : 0.8587659999999999 (48)
 - test.top_1.conv1 : 0.14894 (51)
 - test.top_1.res2 : 0.28634 (51)
 - test.top_1.res3 : 0.39478 (51)
 - test.top_1.res4 : 0.56746 (51)
 **- test.top_1.res5 : 0.6443799999999998 (35)**
 - test.top_5.conv1 : 0.29944 (55)
 - test.top_5.res2 : 0.48829999999999996 (55)
 - test.top_5.res3 : 0.61498 (51)
 - test.top_5.res4 : 0.78732 (55)
 - test.top_5.res5 : 0.86064 (47)
```

I don't believe that results are guaranteed to be the same across the 8gpu and 1gpu schemes here. Note that scaling the lr is based on the Imagenet in 1 hour paper: https://arxiv.org/pdf/1706.02677.pdf. There are some differences with the papers results and these experiments. The paper tests global batch sizes of 256+ -- here the global batch size is 32. Note also the Imagenet in 1 hour paper does not calculate the BN statistics across all worker, whereas these transfer experiments do (By setting CONVERT_BN_TO_SYNC_BN: True).

So imo, I see reproducing the 8gpu numbers in a 1gpu scheme as an area of research. You could start by tuning some of the hyperaparameters -- imo batch size, LR, and weight decay values could be a good place to start. 

(@prigoyal lmk if you disagree / If I've mischaracterized something).

**Comment by prigoyal** [comment]
thank you @iseessel for the debugging and for all the data points above. We expect the results to reproduce between the 8gpu and 1gpu accounting for all the differences you spotted (sync BN etc). 
With that said, I will take this task on further from here, as it's important to understand this further for our research. Thank you for your efforts on this. All above data points are helpful.
